README.md
# SeoTitles
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/Space48/SeoTitles/badges/quality-score.png?b=master&s=f13485e424f50da6599321152195ef4d60e4a456)](https://scrutinizer-ci.com/g/Space48/SeoTitles/?branch=master)
[![Build Status](https://scrutinizer-ci.com/g/Space48/SeoTitles/badges/build.png?b=master&s=48100e53b59c43fb14b14de8039d728546f5ee6d)](https://scrutinizer-ci.com/g/Space48/SeoTitles/build-status/master)
[![Code Coverage](https://scrutinizer-ci.com/g/Space48/SeoTitles/badges/coverage.png?b=master&s=ad73aed196c7f8fbf95cf7a1d8da28f33a8a3f58)](https://scrutinizer-ci.com/g/Space48/SeoTitles/?branch=master)

SeoTitles module for Magento 2

<img width="1249" alt="indoor_living___categories___inventory___products___magento_admin" src="https://cloud.githubusercontent.com/assets/1080386/25120747/37ce3e1a-2417-11e7-9835-a748f7fb5165.png">
This module adds and extra leyer to override H1 values for categories and products.

## Installation

### Manually 

To install this module copy the code from this repository to `app/code/Space48/SeoTitles` folder of your Magento 2 instance,  you need to run the following commands from the terminal:
 
1. `bin/magento setup:upgrade`

### Using Composer

From the terminal execute the following commands:

1. `composer config repositories.space48-seo-titles vcs git@github.com:Space48/SeoTitles.git`
2. `composer require "space48/SeoTitles"` to install the latest version or `"space48/SeoTitles:{module-version}"` to install a specific [version](https://github.com/Space48/seotitles/releases).
3. `bin/magento setup:upgrade`

### Using Modman (Good for development)

From the terminal execute the following:

1. `modman init` (in case Modman already initialized there)
2. `modman clone git@github.com:Space48/SeoTitles.git`
3. `bin/magento setup:upgrade`

## How to use it
Once installed, go to the `Admin Panel -> Products -> Categories -> Select a category`, 
and under `Search Engine Optimisation` section you will see a field called `H1 Override`.


## How to uninstall

From the terminal execute the following:

`bin/magento module:uninstall Space48_SeoTitles`

**Note:** This only works if you have installed the module using composer.

## How to test it
### Steps
1. Go to the `Admin Panel -> Products -> Categories` and select a category, then 
under `Search Engine Optimisation` section you will see a field called `H1 Override`.
2. Write in there `Seo Title Works` and save changes.
![seo_title_test__id__4____categories___inventory___catalog___magento_admin](https://user-images.githubusercontent.com/1080386/35106906-71ec7ea8-fc67-11e7-90bb-a14ebfcbcbaa.jpg)

3. Check changes in the frontend.

### Expected
- You should see the title of the category has changed.
- You should see the name of the category in the breadcrumbs has changed.
- You should see the menu item related to that category has NOT changed.
![seo_title_test](https://user-images.githubusercontent.com/1080386/35107016-bf28ac28-fc67-11e7-8e7f-37b14ee1f923.jpg)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.5] - 2022-06-14

### Added

- Added php 8 support (no code changes required)

## [1.2.4] - 2020-11-30

### Fixed

- fixed PHP version dependency

## [1.2.3] - 2020-11-27

### Changed

- Added support for PHP 7.4

## [1.2.2] - 2019-06-03

### Changed

- Added support for PHP 7.2-7.3
- Added support for Magento 2.3

## [1.2.1] - 2018-02-15

### Changed

- Added support for PHP 7.1

## [1.2.0] - 2018-01-09

### Removed

- Removed dependency on Space48_DBUpdates module

### Added

- Added uninstall for category attribute

## [1.1.1] - 2017-10-31

### Changed

- Added support for Magento 2.2

## [1.1.0] - 2017-05-05

## [1.0.1] - 2017-04-18

## [1.0.0] - 2017-04-18

Initial release


Setup/Uninstall.php
<?php
/**
 * Space48_SeoTitles
 *
 * @category    Space48
 * @package     Space48_SeoTitles
 * @Date        04/2017
 * @license     http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 * @author      @diazwatson
 */

namespace Space48\SeoTitles\Setup;

use Magento\Catalog\Model\Category;
use Magento\Catalog\Model\Product;
use Magento\Eav\Api\AttributeRepositoryInterface;
use Magento\Eav\Api\Data\AttributeInterface;
use Magento\Framework\Setup\ModuleContextInterface;
use Magento\Framework\Setup\SchemaSetupInterface;
use Magento\Framework\Setup\UninstallInterface;
use Magento\Framework\Exception\NoSuchEntityException;
use Magento\Framework\Exception\StateException;

class Uninstall implements UninstallInterface
{

    /**
     * @var AttributeRepositoryInterface
     */
    private $attributeRepository;

    /**
     * Uninstall constructor.
     *
     * @param AttributeRepositoryInterface $attributeRepository
     */
    public function __construct(AttributeRepositoryInterface $attributeRepository)
    {

        $this->attributeRepository = $attributeRepository;
    }

    /**
     * Invoked when remove-data flag is set during module uninstall.
     *
     * @param SchemaSetupInterface $setup
     * @param ModuleContextInterface $context
     *
     * @return void
     * @throws NoSuchEntityException
     * @throws StateException
     */
    public function uninstall(SchemaSetupInterface $setup, ModuleContextInterface $context)
    {
        $entities = [Category::ENTITY, Product::ENTITY];

        foreach ($entities as $entity) {
            if ($attribute = $this->getAttribute($entity)) {
                $this->attributeRepository->delete($attribute);
            }
        }
    }

    /**
     * Get attribute based on entity
     *
     * @param $entity
     * @return AttributeInterface
     * @throws NoSuchEntityException
     */
    private function getAttribute($entity)
    {
        try {
            $attribute = $this->attributeRepository->get($entity, 'h1_override');
        } catch (NoSuchEntityException $e) {
            $attribute = false;
        }
        return $attribute;
    }
}

Setup/InstallData.php
<?php
/**
 * Space48_SeoTitles
 *
 * @category    Space48
 * @package     Space48_SeoTitles
 * @Date        04/2017
 * @license     http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 * @author      @diazwatson
 */

namespace Space48\SeoTitles\Setup;

use Magento\Eav\Model\Entity\Attribute\ScopedAttributeInterface;
use Magento\Framework\Setup\InstallDataInterface;
use Magento\Framework\Setup\ModuleContextInterface;
use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Catalog\Model\Category;
use Magento\Eav\Setup\EavSetupFactory;

class InstallData implements InstallDataInterface
{

    /**
     * @var EavSetupFactory
     */
    private $eavSetupFactory;

    /**
     * UpgradeData constructor.
     *
     * @param EavSetupFactory $eavSetupFactory
     *
     */
    public function __construct(
        EavSetupFactory $eavSetupFactory
    )
    {
        $this->eavSetupFactory = $eavSetupFactory;
    }

    /**
     * Installs data for a module
     *
     * @param ModuleDataSetupInterface $setup
     * @param ModuleContextInterface $context
     *
     * @return void
     */
    public function install(ModuleDataSetupInterface $setup, ModuleContextInterface $context)
    {
        if (version_compare($context->getVersion(), '1.0.0', '<')) {

            /** @var \Magento\Eav\Setup\EavSetup $eavSetup */
            $eavSetup = $this->eavSetupFactory->create(['setup' => $setup]);

            $eavSetup->addAttribute(
                Category::ENTITY,
                'h1_override',
                [
                    'type'       => 'varchar',
                    'label'      => 'H1 Override',
                    'input'      => 'text',
                    'required'   => false,
                    'sort_order' => 100,
                    'global'     => ScopedAttributeInterface::SCOPE_STORE,
                    'group'      => 'General Information',

                ]);
        }
    }
}

Plugin/Catalog/Model/Product.php
<?php
/**
 * Space48_SeoTitles
 *
 * @category    Space48
 * @package     Space48_SeoTitles
 * @Date        04/2017
 * @license     http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 * @author      @diazwatson
 */

namespace Space48\SeoTitles\Plugin\Catalog\Model;

use Magento\Catalog\Model\ResourceModel\Category\CollectionFactory;
use Magento\Store\Model\StoreManagerInterface;

class Product
{

    /**
     * @var CollectionFactory
     */
    private $collectionFactory;

    /**
     * @var StoreManagerInterface
     */
    private $storeManager;

    /**
     * Category constructor.
     *
     * @param CollectionFactory     $collectionFactory
     * @param StoreManagerInterface $storeManager
     */
    public function __construct(
        CollectionFactory $collectionFactory,
        StoreManagerInterface $storeManager
    ) {
        $this->collectionFactory = $collectionFactory;
        $this->storeManager = $storeManager;
    }

    /**
     * @param $product
     * @param $result
     *
     * @return string
     */
    public function afterGetName($product, $result)
    {
        return $product->getData('h1_override') ? $product->getData('h1_override') : $result;
    }

}

Plugin/Catalog/Model/Category.php
<?php
/**
 * Space48_SeoTitles
 *
 * @category    Space48
 * @package     Space48_SeoTitles
 * @Date        04/2017
 * @license     http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 * @author      @diazwatson
 */

namespace Space48\SeoTitles\Plugin\Catalog\Model;

use Magento\Catalog\Model\ResourceModel\Category\CollectionFactory;
use Magento\Store\Model\StoreManagerInterface;

class Category
{

    /**
     * @var CollectionFactory
     */
    private $collectionFactory;
    /**
     * @var StoreManagerInterface
     */
    private $storeManager;

    /**
     * Category constructor.
     *
     * @param CollectionFactory     $collectionFactory
     * @param StoreManagerInterface $storeManager
     */
    public function __construct(
        CollectionFactory $collectionFactory,
        StoreManagerInterface $storeManager
    )
    {
        $this->collectionFactory = $collectionFactory;
        $this->storeManager = $storeManager;
    }

    public function afterGetName($subject, $result)
    {
        return $subject->getData('h1_override') ? $subject->getData('h1_override') : $result;
    }

    public function afterGetParentCategories($subject)
    {
        $pathIds = array_reverse(explode(',', $subject->getPathInStore()));
        /** @var \Magento\Catalog\Model\ResourceModel\Category\Collection $categories */
        $categories = $this->collectionFactory->create();

        return $categories->setStore(
            $this->storeManager->getStore()
        )->addAttributeToSelect(
            'name'
        )->addAttributeToSelect(
            'h1_override'
        )->addAttributeToSelect(
            'url_key'
        )->addFieldToFilter(
            'entity_id',
            ['in' => $pathIds]
        )->addFieldToFilter(
            'is_active',
            1
        )->load()->getItems();
    }
}

Setup/UpgradeData.php
<?php
/**
 * Space48_SeoTitles
 *
 * @category    Space48
 * @package     Space48_SeoTitles
 * @Date        04/2017
 * @license     http://opensource.org/licenses/afl-3.0.php  Academic Free License (AFL 3.0)
 * @author      @diazwatson
 */

namespace space48\seotitles\Setup;

use Magento\Catalog\Model\Product;
use Magento\Framework\Setup\ModuleContextInterface;
use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\UpgradeDataInterface;
use Magento\Eav\Model\Entity\Attribute\ScopedAttributeInterface;
use Magento\Eav\Setup\EavSetupFactory;

class UpgradeData implements UpgradeDataInterface
{

    /**
     * EAV setup factory
     *
     * @var EavSetupFactory
     */
    private $eavSetupFactory;

    /**
     * UpgradeData constructor.
     *
     * @param EavSetupFactory $eavSetupFactory
     */
    public function __construct(
        EavSetupFactory $eavSetupFactory
    )
    {
        $this->eavSetupFactory = $eavSetupFactory;
    }

    /**
     * Add Override Attribute to Product Entity
     *
     * @param ModuleDataSetupInterface $setup
     * @param ModuleContextInterface $context
     *
     * @return void
     */
    public function upgrade(ModuleDataSetupInterface $setup, ModuleContextInterface $context)
    {

        if (version_compare($context->getVersion(), '1.1.0', '<')) {

            /** @var \Magento\Eav\Setup\EavSetup $eavSetup */
            $eavSetup = $this->eavSetupFactory->create(['setup' => $setup]);

            $eavSetup->addAttribute(Product::ENTITY, 'h1_override',
                [
                    'type'                    => 'varchar',
                    'backend'                 => '',
                    'frontend'                => '',
                    'label'                   => 'H1 Override',
                    'input'                   => 'text',
                    'class'                   => '',
                    'global'                  => ScopedAttributeInterface::SCOPE_STORE,
                    'visible'                 => true,
                    'required'                => false,
                    'user_defined'            => false,
                    'default'                 => null,
                    'nullable'                => true,
                    'searchable'              => false,
                    'filterable'              => false,
                    'comparable'              => false,
                    'visible_on_front'        => false,
                    'used_in_product_listing' => false,
                    'unique'                  => false,
                ]
            );
        }

    }
}